framework/src/transfers/MultiAppMeshFunctionTransfer.C

```diff
@@ -63,289 +62,347 @@ MultiAppMeshFunctionTransfer::execute()
-      else
-        tmp_to_mesh = &to_problem.mesh().getMesh();
+  ////////////////////
+  // Gather all of the evaluations, pick out the best ones for each point, and
+  // apply them to the solution vector.  When we are transferring from
+  // multiapps, there may be multiple overlapping apps for a particular point.
+  // In that case, we'll try to use the value from the app with the lowest id.
+  ////////////////////
+  for (processor_id_type i_proc = 0; i_proc < n_processors(); i_proc++)
+  {
+    if (i_proc == processor_id())
+      continue;
 
-      MeshBase & to_mesh = *tmp_to_mesh;
+    _communicator.receive(i_proc, incoming_evals[i_proc]);
+    if (_direction == FROM_MULTIAPP)
+      _communicator.receive(i_proc, incoming_app_ids[i_proc]);
+  }
 
-      bool is_nodal = to_sys.variable_type(to_var_num).family == LAGRANGE;
+  for (unsigned int i_to = 0; i_to < _to_problems.size(); i_to++)
+  {
+    System * to_sys = find_sys(*_to_es[i_to], _to_var_name);
 
-      for (unsigned int i=0; i<_multi_app->numGlobalApps(); i++)
-      {
-        if (!_multi_app->hasLocalApp(i))
-          continue;
+    unsigned int sys_num = to_sys->number();
+    unsigned int var_num = to_sys->variable_number(_to_var_name);
 
-        MPI_Comm swapped = Moose::swapLibMeshComm(_multi_app->comm());
-        FEProblem & from_problem = *_multi_app->appProblem(i);
-        MooseVariable & from_var = from_problem.getVariable(0, _from_var_name);
-        SystemBase & from_system_base = from_var.sys();
+    NumericVector<Real> * solution;
+    switch (_direction)
+    {
+      case TO_MULTIAPP:
+        solution = & getTransferVector(i_to, _to_var_name);
+        break;
+      case FROM_MULTIAPP:
+        solution = to_sys->solution.get();
+        break;
+    }
 
-        System & from_sys = from_system_base.system();
+    MeshBase * to_mesh = & _to_meshes[i_to]->getMesh();
 
-        // Only works with a serialized mesh to transfer from!
-        mooseAssert(from_sys.get_mesh().is_serial(), "MultiAppMeshFunctionTransfer only works with SerialMesh!");
+    bool is_nodal = to_sys->variable_type(var_num).family == LAGRANGE;
 
-        unsigned int from_var_num = from_sys.variable_number(from_var.name());
+    if (is_nodal)
+    {
+      MeshBase::const_node_iterator node_it = to_mesh->local_nodes_begin();
+      MeshBase::const_node_iterator node_end = to_mesh->local_nodes_end();
 
-        EquationSystems * tmp_es = NULL;
+      for (; node_it != node_end; ++node_it)
+      {
+        Node * node = *node_it;
 
-        if (_displaced_source_mesh && from_problem.getDisplacedProblem())
-          tmp_es = &from_problem.getDisplacedProblem()->es();
-        else
-          tmp_es = &from_problem.es();
+        // Skip this node if the variable has no dofs at it.
+        if (node->n_dofs(sys_num, var_num) < 1)
+          continue;
 
-        EquationSystems & from_es = *tmp_es;
+        unsigned int lowest_app_rank = -1; // -1 = largest unsigned int
```

Comment:
For consistency, please use `libMesh::invalid_uint` instead of -1, even though they are the same thing.